pkg/resourcecollector/resourcecollector.go

```diff
@@ -0,0 +1,378 @@
+package resourcecollector
+
+import (
+	"fmt"
+
+	"github.com/heptio/ark/pkg/discovery"
+	"github.com/heptio/ark/pkg/util/collections"
+	"github.com/libopenstorage/stork/drivers/volume"
+	"github.com/portworx/sched-ops/k8s"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+)
+
+// ResourceCollector is used to collect and process unstructured objects in namespaces and using label selectors
+type ResourceCollector struct {
+	Driver           volume.Driver
+	discoveryHelper  discovery.Helper
+	dynamicInterface dynamic.Interface
+}
+
+// Init initializes the resource collector
+func (r *ResourceCollector) Init() error {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return fmt.Errorf("error getting cluster config: %v", err)
+	}
+
+	aeclient, err := apiextensionsclient.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("error getting apiextention client, %v", err)
+	}
+
+	discoveryClient := aeclient.Discovery()
+	r.discoveryHelper, err = discovery.NewHelper(discoveryClient, logrus.New())
+	if err != nil {
+		return err
+	}
+	err = r.discoveryHelper.Refresh()
+	if err != nil {
+		return err
+	}
+	r.dynamicInterface, err = dynamic.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func resourceToBeCollected(resource metav1.APIResource) bool {
+	// Deployment is present in "apps" and "extensions" group, so ignore
+	// "extensions"
+	if resource.Group == "extensions" && resource.Kind == "Deployment" {
+		return false
+	}
+
+	switch resource.Kind {
+	case "PersistentVolumeClaim",
```

Comment:
we should allow users to configure this list as this CRDs, we would skip those leading to a partial migration.

Also I think we discussed this before but this needs to have storage classes. A statefulset scale up will fail post migration, which means we aren't doing a complete statefulset migration.

Reply:
There is already an open issue for migrating CRs and CRDs (#282). There is more effort required to support that than just allowing them here since the resources created by the CRs don't need to be migrated themselves. Also, need to be able to handle cases where the CRD isn't registered on the destination.

Storageclasses are cluster specific, admins will create them during cluster install. Although a good to have I don't see that as a blocker for anything.